hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/io/FlinkMergeHandle.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.io;
+
+import org.apache.hudi.client.WriteStatus;
+import org.apache.hudi.common.engine.TaskContextSupplier;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.HoodieBaseFile;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.hudi.common.util.DefaultSizeEstimator;
+import org.apache.hudi.common.util.HoodieRecordSizeEstimator;
+import org.apache.hudi.common.util.HoodieTimer;
+import org.apache.hudi.common.util.collection.ExternalSpillableMap;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.table.HoodieTable;
+
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link HoodieMergeHandle} that supports merge write incrementally(mini-batches).
+ *
+ * <p>For a new mini-batch, it initialize and set up the next file path to write,
+ * and closes the file path when the mini-batch write finish. When next mini-batch
+ * write starts, it rolls over to another new file. If all the mini-batches write finish
+ * for a checkpoint round, it renames the last new file path as the desired file name
+ * (name with the expected file ID).
+ *
```

Comment:
Sorry, I am a bit confused by mini-batch, checkpoint, and fileid e.t.c. 
We did a shuffle by file id before the `bucketWriter`, so each `bucketWriter` should handle one file id in each commit right? Do you mean we will write multiple files for one commit(checkpoint) in a single `bucketWriter`?

Reply:
The `BucketWriter` hold several write handles, each for a bucket, during one checkpoint, we may write multiple times(as mini-batches) for one bucket, e.g. we may append to the bucket handle as multiple times, for both create handle and merge handle, based on how mini-batches there are in one checkpoint for the `BucketWriter`.

Reply:
I mean can we remove the `MiniBatchHandle ` interface?

Reply:
No, a new interface make the handle more clear because it is different with spark style handle.

Reply:
I know your thought. But from another view, does it the only way for flink, and will you implement another way?
If we only provide one way like this, we may not need to define it. It will increase the cost of understanding like gray's thought.